src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregatorStore.cs

```diff
@@ -0,0 +1,472 @@
+    //   the fast path.
+    //   - Allocations per lookup: Any lookup of 3 or fewer labels on the above fast path is allocation free. We have separate
+    //   dictionaries dependending on the number of labels in the list and the dictionary keys are structures representing fixed size
+    //   lists of strings or objects. For example with two labels the lookup is done in a
+    //   FixedSizeLabelNameDictionary<StringSequence2, ConcurrentDictionary<ObjectSequence2, TAggregator>>
+    //   Above 3 labels we have StringSequenceMany and ObjectSequenceMany which wraps an underlying string[] or object?[] respectively.
+    //   Doing a lookup with those types will need to do allocations for those arrays.
+    //   - Total memory footprint per-store: We have a store for every instrument we are tracking and an entry in the 2nd level
+    //   dictionary for every label set. This can add up to a lot of entries. Splitting the label sets into keys and values means we
+    //   only need to store each unique key list once (as the key of the 1st level dictionary). It is common for all labelsets on an
+    //   instrument to have the same keys so this can be a sizable savings. We also use a union to store the 1st level dictionaries
+    //   for different label set sizes because most instruments always specify labelsets with the same number of labels (most likely
+    //   zero).
+    [System.Security.SecuritySafeCritical] // using SecurityCritical type ReadOnlySpan
+    internal struct AggregatorStore<TAggregator> where TAggregator : Aggregator
+    {
+        // this union can be:
+        // null
+        // TAggregator
+        // FixedSizeLabelNameDictionary<StringSequence1, ConcurrentDictionary<ObjectSequence1, TAggregator>>
+        // FixedSizeLabelNameDictionary<StringSequence2, ConcurrentDictionary<ObjectSequence2, TAggregator>>
+        // FixedSizeLabelNameDictionary<StringSequence3, ConcurrentDictionary<ObjectSequence3, TAggregator>>
+        // FixedSizeLabelNameDictionary<StringSequenceMany, ConcurrentDictionary<ObjectSequenceMany, TAggregator>>
+        // MultiSizeLabelNameDictionary<TAggregator> - this is used when we need to store more than one of the above union items
+        private volatile object? _stateUnion;
+        private volatile AggregatorLookupFunc<TAggregator>? _cachedLookupFunc;
+        private Func<TAggregator> _createAggregatorFunc;
+
+        public AggregatorStore(Func<TAggregator> createAggregator)
+        {
+            _stateUnion = default;
+            _cachedLookupFunc = default;
+            _createAggregatorFunc = createAggregator;
+        }
+
+        public TAggregator GetAggregator(ReadOnlySpan<KeyValuePair<string, object?>> labels)
+        {
+            AggregatorLookupFunc<TAggregator>? lookupFunc = _cachedLookupFunc;
+            if (lookupFunc != null)
+            {
+                if (lookupFunc(labels, out TAggregator? aggregator)) return aggregator!;
```

Comment:
This `!` shouldn't be necessary if you change the definition of the `AggregatorLookupFunc` delegate to have a `[NotNullWhen(true)]` attribute on the out argument: `[NotNullWhen(true)] out TAggregator? aggregator`.